hnn_core/tests/test_network.py

```diff
@@ -146,22 +146,156 @@ def test_network():
 
     # create a new connection between cell types
+    net = Network(deepcopy(params), add_drives_from_params=True)
+    src_gids = net.gid_ranges['bursty1']
+    src_types = np.repeat('bursty1', len(src_gids))
+    target_gids = net.gid_ranges['L5_basket']
+    target_types = np.repeat('L5_basket', len(target_gids))
+
     nc_dict = {'A_delay': 1, 'A_weight': 1e-5, 'lamtha': 20,
                'threshold': 0.5}
-    network_builder._connect_celltypes(
-        'bursty1', 'L5Basket', 'soma', 'gabaa', nc_dict,
-        unique=False)
+    connectivity = net._all_to_all_connect(
+        src_types, src_gids, target_types, target_gids,
+        'soma', 'gabaa', nc_dict, unique=False)
+    net.connectivity_list.extend(connectivity)
+    network_builder = NetworkBuilder(net)
     assert 'bursty1_L5Basket_gabaa' in network_builder.ncs
     n_conn = len(net.gid_ranges['bursty1']) * len(net.gid_ranges['L5_basket'])
     assert len(network_builder.ncs['bursty1_L5Basket_gabaa']) == n_conn
 
     # try unique=True
-    network_builder._connect_celltypes(
-        'extgauss', 'L5Basket', 'soma', 'gabaa', nc_dict,
-        unique=True)
+    net = Network(deepcopy(params), add_drives_from_params=True)
+    src_gids = net.gid_ranges['extgauss']
+    src_types = np.repeat('extgauss', len(src_gids))
+    connectivity = net._all_to_all_connect(
+        src_types, src_gids, target_types, target_gids,
+        'soma', 'gabaa', nc_dict, unique=True)
+    net.connectivity_list.extend(connectivity)
+    network_builder = NetworkBuilder(net)
     n_conn = len(net.gid_ranges['L5_basket'])
     assert len(network_builder.ncs['extgauss_L5Basket_gabaa']) == n_conn
 
+    # Test inputs for connectivity API
+    connection = {
+        'src_type': 'L2Pyr', 'src_gid': 36, 'target_type': 'L2Pyr',
+        'target_gid': 35, 'loc': 'proximal', 'receptor': 'nmda',
+        'nc_dict': {
+            'A_delay': 1.0, 'threshold': 0.0, 'lamtha': 3.0,
+            'A_weight': 0.0005}}
+    net.add_connection(connection)
+
+    with pytest.raises(
+            TypeError, match='connectivity must be one of dict '
+            'or list of dict, got float'):
+        net.add_connection(1.0)
+
+    test_connection = deepcopy(connection)
+    test_connection['src_gid'] = 1.0
+    with pytest.raises(
+            TypeError, match=r"connectivity\['src_gid'\] must be "
+            r"of type int, got float"):
+        net.add_connection(test_connection)
+
+    test_connection = deepcopy(connection)
+    test_connection['src_type'] = 1.0
+    with pytest.raises(
+            TypeError, match=r"connectivity\['src_type'\] must be "
+            r"of type str, got float"):
+        net.add_connection(test_connection)
+
+    test_connection = deepcopy(connection)
+    test_connection['target_gid'] = 1.0
+    with pytest.raises(
+            TypeError, match=r"connectivity\['target_gid'\] must be "
+            r"of type int, got float"):
+        net.add_connection(test_connection)
```

Comment:
the tests are taking up too much real estate, you can just do for loops etc. 

But anyhow I'd worry about the tests later. My immediate concern is I want to see a realistic example how you'd use it. Can you update an existing example to show me?